### PR TITLE
Login backend

### PR DESCRIPTION
**Pull request by xleonx0x**
improved everything / still no session auth added so hold off

### CODE REVIEW

**Comment by lachlanshoesmith** [comment]
> so hold off

ok 👍

File: backend/src/index.ts [hunk 1]
```diff
@@ -1,16 +1,64 @@
...
+  const session = await getDocs(sessionData);
+
+  if (session.empty) {
+    return false;
+  }
+
+  // if current time is greater than expiration date, return false and
+  // removes it from the database
+  if (new Date() > session.docs[0].data().expirationDate.toDate()) {
+    console.log(new Date(), session.docs[0].data().expirationDate);
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: this can be removed now

File: backend/src/index.ts [hunk 2]
```diff
@@ -1,16 +1,64 @@
...
+    await session_remove(sessionId);
+    return false;
+  }
+  return true;
+};
+
+// Clears all expired sessions in the db
+// Idea, periodically call this to stay efficient on document reads?
+const clean_sessions = async (): Promise<void> => {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: this function doesnt seem to really take advantage of returning a promise

File: backend/src/index.ts [hunk 3]
```diff
@@ -1,16 +1,64 @@
...
+  where,
+  query,
+  deleteDoc,
+} from "firebase/firestore";
 import { db } from "./firebase";
 import cors from "cors";
 import crypto from "crypto";
 
-const EXPRESS_PORT = 3000;
+const session_auth = async (sessionId: string): Promise<boolean> => {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: this function doesn't seem to return a promise but rather just a boolean?

File: backend/src/index.ts [hunk 4]
```diff
@@ -1,16 +1,64 @@
...
+  const allSessions = await getDocs(sessions);
+  allSessions.forEach(async (sessionDoc) => {
+    if (new Date() > sessionDoc.data().expirationDate.toDate()) {
+      const ref = sessionDoc.ref;
+      await deleteDoc(ref);
+    }
+  });
+};
+
+const session_remove = async (sessionId: string): Promise<void> => {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: returns nothing

**Review by lachlanshoesmith — CHANGES_REQUESTED** [review]

File: backend/src/index.ts (hunk 2 above)
Comment (xleonx0x) [reply backend/src/index.ts]: so a func can be async without containing a promise?

File: backend/src/index.ts (hunk 3 above)
Comment (xleonx0x) [reply backend/src/index.ts]: it says this 
> The return type of an async function or method must be the global Promise<T> type. Did you mean to write 'Promise<boolean>'?ts(1064)

so I'm unsure what to put

Comment (xleonx0x) [reply backend/src/index.ts]: i just removed the return types from them entirely 👍

File: backend/src/index.ts (hunk 2 above)
Comment (lachlanshoesmith) [reply backend/src/index.ts]: sorry what i mean is you should be able to get away with not specifying the return type here

Comment (lachlanshoesmith) [reply backend/src/index.ts]: if the linter insists u must specify that then disregard, keep as is - my b

File: backend/src/index.ts (hunk 3 above)
Comment (lachlanshoesmith) [reply backend/src/index.ts]: if it doesnt get mad then yeah that sounds good

**Review by lachlanshoesmith — DISMISSED** [review]

**Comment by xleonx0x** [comment]
i have no idea if this works with frontend

File: backend/src/index.ts [hunk 5]
```diff
@@ -26,32 +80,42 @@ app.use(
     },
     secret: process.env.SESSION_SECRET as string,
     saveUninitialized: false,
-    resave: true
+    resave: true,
   }),
 );
 
-app.post('/logout', async (req, res) => {
+app.post("/session/validate", async (req: Request, res: Response) => {
```
Comment (bengodw) [inline backend/src/index.ts]: is there a reason why this needs to be an endpoint - can this just be a function that gets called at the start of each endpoint that requires a valid session

Comment (xleonx0x) [reply backend/src/index.ts]: I was thinking that the frontend needs to change the navbar e.g login / logout based on if the person is logged in, so I was gonna call this each time to change stuff like navbar / profile dropdown. 🔒 based on if the person is a guest or user

File: backend/src/interfaces.ts [hunk 6]
```diff
@@ -22,15 +24,14 @@ export interface Game {
...
-  id: string;
   username: string;
   password: string;
   salt: string;
   profilePicture?: string;
   highScore?: number;
   cumulativeScore?: number;
   shirts?: number; // Would this refer to the number of secrets(clothes drawing) found?
-  dateJoined?: Date;
+  dateJoined: Date | Timestamp;
```
Comment (bengodw) [inline backend/src/interfaces.ts:34]: probably should be one or the other

File: backend/src/index.ts [hunk 7]
```diff
@@ -1,16 +1,70 @@
...
+  }
+  return true;
+};
+
+// Clears all expired sessions in the db
+// Idea, periodically call this to stay efficient on document reads?
+const clean_sessions = async () => {
+  const sessions = collection(db, "sessions");
+  const allSessions = await getDocs(sessions);
+  allSessions.forEach(async (sessionDoc) => {
```
Comment (bengodw) [inline backend/src/index.ts]: why doesn't this use session_remove?

Comment (bengodw) [reply backend/src/index.ts]: also doesn't this and session_auth do the same thing

File: backend/src/index.ts [hunk 8]
```diff
@@ -26,32 +80,42 @@ app.use(
...
-    console.log('cookies removed');
-  });
+  const userId = session.docs[0].data().userId;
+  const userDocRef = doc(db, "users", userId);
 
-  res.send('Goodbye!').status(200);
-})
+  getDoc(userDocRef).then((docSnap) => {
+    const data = docSnap.data() as User;
+    const { password, salt, ...sanitizedData } = data;
```
Comment (bengodw) [inline backend/src/index.ts:93]: satisfying use of spread operator

File: backend/src/index.ts (hunk 5 above)
Comment (xleonx0x) [reply backend/src/index.ts]: actually your right i have a user end point for that - 😔 ill remove it

**Review by bengodw — COMMENTED** [review]
not a complete review - will take a closer look at the session implementation later when I get a chance

File: backend/src/interfaces.ts (hunk 6 above)
Comment (xleonx0x) [reply backend/src/interfaces.ts:34]: ok so, firebase converts the Date to a Timestamp so I have that so I can convert it back when I return the user info without it complaining

File: backend/src/index.ts (hunk 7 above)
Comment (xleonx0x) [reply backend/src/index.ts]: i was just thinking about situations where the session db will get bigger and bigger but yea we can deal with that later I'll delete that

File: backend/src/index.ts [hunk 9]
```diff
@@ -72,43 +117,84 @@ app.post("/register", async (req: TypedRequest<LoginBody>, res: Response) => {
...
-      return result
-        ? res.status(201).send("User Successfully Logged In")
-        : res.status(401).send("Incorrect password");
+      if (result) {
+        req.session.regenerate(async (err: Error | null) => {
+          if (err) {
+            return res.status(500).send("Error regenerating session.");
+          }
+          const expiryTime: Date = new Date();
+          expiryTime.setDate(expiryTime.getDate() + 7);
```
Comment (bengodw) [inline backend/src/index.ts:167]: not sure if this will work

**Review by bengodw — APPROVED** [review]
can't rly see any problems - maybe in future would make sense to integrate express-session's store feature: https://www.npmjs.com/package/@google-cloud/connect-firestore

**Pull request merged**